### ISSUE TITLE
Have sumvalue accept higher-order arrays

### ISSUE THREAD

**Issue by ahwillia** [JuliaML/Losses.jl#28]
For matrix decompositions, factor analysis, similar models the predicted object is a matrix or higher-order array of data. In this case, the total loss could be calculated 

```julia
sumvalue(loss::Loss, target::Matrix, output::Matrix)
```

As mentioned in passing in https://github.com/JuliaML/Losses.jl/issues/27, currently `sumvalue` requires `target::AbstractVector` which forces me to vectorize `target` before calling the function:

```julia
sumvalue(loss, view(target,:), view(output,:))
```

This is a fine workaround for now.

But my preference would be to have `sumvalue` accept `target` and `output` as `AbstractArrays` and check that they have the same dimensions (throwing an error if the check fails). Then it iterates over the two arrays and computes the total loss.

**Comment by Evizero** [comment]
so I assume in your usecase element (i,j) from target is compared to element (i,j) from output and the two arrays need to be of the same structure?

**Comment by ahwillia** [comment]
Yes, exactly.

I would also like to be able to do `value(loss, target::AbstractArray{T,N}, output::AbstractArray{T,N})` which returns a new array with the loss taken elementwise.

**Comment by tbreloff** [comment]
This is exactly the type of definition that I'd like to automatically build with the experimental stuff in Transformations.  Basically I'd like to write:

```julia
@loss L1DistLoss(target, output) = sum(abs2.(target .- output))
```

And all of the boilerplate stuff would be unnecessary... it would allow for many dimensions, and would know the properties of the output given the components of the computation, as well as the ability to compute derivatives analytically, etc.  In my head, the macro creates a "Loss Factory" of sorts, which can produce custom code given the input/output shape, learning algorithm, or anything else that might be decided dynamically.

Having said all that, I'm still confused on the best approach to make it happen!

**Comment by Evizero** [comment]
I agree with you there, but for typically losses we should imho still handcraft the most optimal implementation. Surely would be nice to offer an interface that allows users to create new losses where the derivatives build automatically. and also to offer a symbolic interface for the losses we do have, so that we can chain stuff together and use julias metaprogramming to compile special stuff on the fly. (and maybe far down the road GPU kernels too)

Here the cool part: I just discovered that someone is working on the heavy lifting for the symbolic stuff already. Check this out https://github.com/dfdx/Espresso.jl

**Comment by Evizero** [comment]
We should try to help @dfdx out where we can. Especially for transformations it would be nice to try and make use of this kind of thing for code generation.

**Comment by ahwillia** [comment]
I'd like to take a crack at this soon (necessary for a project I'm working on). The solution might be as simple as defining some fallback functions.

```julia
function value!(buffer::AbstractArray, loss::SupervisedLoss, target::AbstractArray, output::AbstractArray)
    value!(view(buffer,:), loss, view(target,:), view(output,:))
end

function deriv!(buffer::AbstractArray, loss::SupervisedLoss, target::AbstractArray, output::AbstractArray)
    deriv!(view(buffer,:), loss, view(target,:), view(output,:))
end

function sumvalue(loss::SupervisedLoss, target::AbstractArray, output::AbstractArray)
    sumvalue(loss, view(target,:), view(output,:))
end

function sumderiv(loss::SupervisedLoss, target::AbstractArray, output::AbstractArray)
    sumderiv(loss, view(target,:), view(output,:))
end
```

Anyone see a reason this is a bad idea? I don't think making the vectorized view would add much overhead. Maybe I should add some benchmarks to confirm this.

**Comment by Evizero** [comment]
Well, I don't know. Have to think about this a bit to give useful feedback. The thing with loss functions is that they are essentially computing kernels of sort that are called in every iteration (so small stuff matters)

**Comment by ahwillia** [comment]
Closed by #38
